src/identity/manager.rs

```diff
@@ -32,16 +36,68 @@ impl fmt::Display for Identity {
 }
 
 impl SecretManager {
     pub fn new(cfg: crate::config::Config) -> SecretManager {
         let client = CaClient::new(cfg.auth);
-        SecretManager { client }
+        let cache: HashMap<Identity, tls::Certs> = Default::default();
+        return SecretManager { client: client, cache: Arc::new(Mutex::new(cache)) };
+    }
+
+    pub async fn refresh_handler(id: Identity, ctx: SecretManager, initial_sleep_time: u64) {
+        info!("refreshing certs for id {} in {:?} seconds", id, initial_sleep_time);
+        sleep(Duration::from_secs(initial_sleep_time)).await;
+        loop {
+            match ctx.client.clone().fetch_certificate(id.clone()).await {
+                Err(e) => {
+                    // Cert refresh has failed. Drop cert from the cache.
+                    warn!("Failed cert refresh for id {:?}: {:?}", id, e);
+                    { // lock cache
+                        let mut locked_cache = ctx.cache.lock().unwrap();
+                        locked_cache.remove(&id.clone());
+                    } // unlock cache
+                    return;
+                }
+                Ok(fetched_certs) => {
+                    info!("refreshed certs {:?}", fetched_certs);
+                    { // lock cache
+                        let mut locked_cache = ctx.cache.lock().unwrap();
+                        locked_cache.insert(id.clone(), fetched_certs.clone());
+                    } // unlock cache
+                    let sleep_dur = Duration::from_secs(fetched_certs.get_seconds_until_refresh());
+                    info!("refreshing certs for id {} in {:?} seconds", id, sleep_dur);
+                    sleep(sleep_dur).await;
+                }
+            }
+        }
     }
 
     #[instrument(skip_all, fields(%id))]
-    pub async fn fetch_certificate(&self, id: Identity) -> Result<tls::Certs, Error> {
-        self.client.clone().fetch_certificate(id).await
+    pub async fn fetch_certificate(&mut self, id: &Identity) -> Result<tls::Certs, Error> {
+        // Check cache first
+        { // lock cache
+            let locked_cache = self.cache.lock().unwrap();
+            let cache_certs: std::option::Option<&tls::Certs> = locked_cache.get(id);
+            info!("cache certs for req: {:?}", cache_certs);
+            if cache_certs.is_some() {
+                return Ok(cache_certs.unwrap().clone())
+            }
+        } // unlock cache
+
+        // No cache entry, fetch it and spawn refresh handler
+        let fetched_certs = self.client.clone().fetch_certificate(id.clone()).await?;
+        info!("fetched certs {:?}", fetched_certs);
+        { // lock cache
+            let mut locked_cache = self.cache.lock().unwrap();
+            locked_cache.insert(id.clone(), fetched_certs.clone());
+        } // unlock cache
+
+        tokio::spawn(SecretManager::refresh_handler(id.clone(),
+                                                        self.clone(),
```

Comment:
the `self.clone` feels odd to me. if we add some Clone field to SecretManager an unsuspecting dev may assume it's a shared reference when it's not. Should we pass a client and reference to the cache here explicitly? Should the cache get wrapped up in some accessors?

I'm not sure of the idiomatic thing to do.